workitem/link/link_repository.go

```diff
@@ -257,9 +303,37 @@ func (r *GormWorkItemLinkRepository) Save(ctx context.Context, linkToSave WorkIt
 		return nil, errors.NewVersionConflictError("version conflict")
 	}
 	linkToSave.Version = linkToSave.Version + 1
-	if err := r.ValidateCorrectSourceAndTargetType(ctx, existingLink.SourceID, existingLink.TargetID, existingLink.LinkTypeID); err != nil {
+
+	linkTypeToSave, err := r.workItemLinkTypeRepo.Load(ctx, linkToSave.LinkTypeID)
+	if err != nil {
+		return nil, errs.Wrap(err, "failed to load link type")
+	}
+
+	if err := r.ValidateCorrectSourceAndTargetType(ctx, linkToSave.SourceID, linkToSave.TargetID, linkTypeToSave.ID); err != nil {
 		return nil, errs.WithStack(err)
 	}
+
+	// check to disallow multiple parents in tree topology
+	if linkTypeToSave.Topology == TopologyTree {
```

Comment:
this (line 140-159 and 316-335) looks a lot like code duplication and I'd outsource it to a separate function to be used from multiple places.